src/config.rs

```diff
@@ -180,12 +212,154 @@ pub fn parse_config() -> Result<Config, Error> {
+                .try_into()
+                .expect("concurrency cannot be negative"),
+        )?,
 
         proxy_args: parse_args(),
     })
 }
+
+#[derive(serde::Deserialize, Default, Clone, PartialEq)]
+#[serde(rename_all = "camelCase")]
+pub struct MeshConfig {
+    pub default_config: Option<ProxyConfig>,
+}
+
+#[derive(serde::Deserialize, Default, Clone, PartialEq)]
+#[serde(rename_all = "camelCase")]
+pub struct ProxyConfig {
+    pub discovery_address: Option<String>,
+    pub proxy_admin_port: Option<u16>,
+    pub status_port: Option<u16>,
+    pub concurrency: ::core::option::Option<u16>,
+    pub termination_drain_duration: Option<time::Duration>,
+    pub proxy_metadata: std::collections::HashMap<String, String>,
+}
+
+impl ProxyConfig {
+    fn merge(mut self, other: Self) -> Self {
+        self.discovery_address = other
+            .discovery_address
+            .or_else(|| self.discovery_address.clone());
+        self.proxy_admin_port = other.proxy_admin_port.or(self.proxy_admin_port);
+        self.status_port = other.status_port.or(self.status_port);
+        self.concurrency = other.concurrency.or(self.concurrency);
+        self.proxy_metadata.extend(other.proxy_metadata);
+        self.termination_drain_duration = other
+            .termination_drain_duration
+            .or(self.termination_drain_duration);
+        self
+    }
+}
+
+fn construct_proxy_config(mc_path: &str, pc_env: Option<&str>) -> anyhow::Result<ProxyConfig> {
+    let mesh_config = match fs::File::open(mc_path) {
+        Ok(f) => serde_yaml::from_reader(f)
+            .map(|v: MeshConfig| v.default_config)
+            .map_err(anyhow::Error::new),
+        Err(e) => {
+            if e.kind() == std::io::ErrorKind::NotFound {
+                Ok(None)
+            } else {
+                Err(anyhow!(e))
+            }
+        }
+    }
+    .map_err(|e| anyhow!("failed parsing mesh config file {}: {}", mc_path, e))?;
+
+    let proxy_config_env = pc_env
+        .map(|pc_env| {
+            if pc_env.is_empty() {
+                Ok(None)
+            } else {
+                serde_yaml::from_str(pc_env)
+            }
+        })
+        .unwrap_or(Ok(None))
+        .map_err(|e| anyhow!("failed parsing proxy config env: {}", e))?;
+
+    let mut pc = [mesh_config, proxy_config_env]
+        .into_iter()
+        .flatten()
+        .fold(ProxyConfig::default(), |pc, v| pc.merge(v));
+
+    // only include ISTIO_META_ prefixed fields in this map
+    // TODO we could use any other items here for the various env vars for construct_config?
+    // https://istio.io/latest/docs/reference/config/istio.mesh.v1alpha1/#:~:text=Additional%20environment%20variables
+    pc.proxy_metadata = pc
+        .proxy_metadata
+        .into_iter()
+        .filter_map(|(k, v)| Some((k.strip_prefix("")?.to_string(), v)))
+        .collect();
+    
+    // TODO if certain fields like trustDomainAliases are added, make sure they merge like:
```

Comment:
fwiw this one is in mesh config. But we may need this and trust_domain eventually actually. Or send it over XDS